{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook roughly investigates if Factor Analysis can correctly find genes in the k = 1 pathway that **DICER1** participates in."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing the data for analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading main gene dependency data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>A1BG (1)</th>\n",
       "      <th>A1CF (29974)</th>\n",
       "      <th>A2M (2)</th>\n",
       "      <th>A2ML1 (144568)</th>\n",
       "      <th>A3GALT2 (127550)</th>\n",
       "      <th>A4GALT (53947)</th>\n",
       "      <th>A4GNT (51146)</th>\n",
       "      <th>AAAS (8086)</th>\n",
       "      <th>AACS (65985)</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH (55055)</th>\n",
       "      <th>ZWINT (11130)</th>\n",
       "      <th>ZXDA (7789)</th>\n",
       "      <th>ZXDB (158586)</th>\n",
       "      <th>ZXDC (79364)</th>\n",
       "      <th>ZYG11A (440590)</th>\n",
       "      <th>ZYG11B (79699)</th>\n",
       "      <th>ZYX (7791)</th>\n",
       "      <th>ZZEF1 (23140)</th>\n",
       "      <th>ZZZ3 (26009)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>0.181332</td>\n",
       "      <td>0.089101</td>\n",
       "      <td>-0.193867</td>\n",
       "      <td>-0.024587</td>\n",
       "      <td>0.038458</td>\n",
       "      <td>-0.181824</td>\n",
       "      <td>0.351672</td>\n",
       "      <td>-0.440073</td>\n",
       "      <td>0.292582</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.124218</td>\n",
       "      <td>-0.469306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.257361</td>\n",
       "      <td>0.244834</td>\n",
       "      <td>-0.408512</td>\n",
       "      <td>0.284734</td>\n",
       "      <td>0.226412</td>\n",
       "      <td>-0.149566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>-0.076383</td>\n",
       "      <td>0.245190</td>\n",
       "      <td>0.191238</td>\n",
       "      <td>0.153008</td>\n",
       "      <td>-0.197035</td>\n",
       "      <td>-0.323295</td>\n",
       "      <td>0.252522</td>\n",
       "      <td>-0.571498</td>\n",
       "      <td>-0.068945</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.212442</td>\n",
       "      <td>-0.426151</td>\n",
       "      <td>-0.068295</td>\n",
       "      <td>0.031635</td>\n",
       "      <td>0.205469</td>\n",
       "      <td>-0.068536</td>\n",
       "      <td>-0.092858</td>\n",
       "      <td>0.074640</td>\n",
       "      <td>0.028779</td>\n",
       "      <td>-0.264840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>0.102195</td>\n",
       "      <td>0.092449</td>\n",
       "      <td>-0.045926</td>\n",
       "      <td>0.171892</td>\n",
       "      <td>0.140561</td>\n",
       "      <td>0.170971</td>\n",
       "      <td>0.086060</td>\n",
       "      <td>-0.432320</td>\n",
       "      <td>0.010664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.083183</td>\n",
       "      <td>-0.269196</td>\n",
       "      <td>0.101146</td>\n",
       "      <td>0.277820</td>\n",
       "      <td>0.208814</td>\n",
       "      <td>0.001393</td>\n",
       "      <td>-0.327514</td>\n",
       "      <td>0.048714</td>\n",
       "      <td>-0.372854</td>\n",
       "      <td>-0.433157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000009</td>\n",
       "      <td>0.142342</td>\n",
       "      <td>-0.033126</td>\n",
       "      <td>-0.051224</td>\n",
       "      <td>0.060560</td>\n",
       "      <td>0.116002</td>\n",
       "      <td>-0.010044</td>\n",
       "      <td>0.104725</td>\n",
       "      <td>-0.610481</td>\n",
       "      <td>0.181508</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.277264</td>\n",
       "      <td>-0.307018</td>\n",
       "      <td>0.044741</td>\n",
       "      <td>0.201551</td>\n",
       "      <td>0.083866</td>\n",
       "      <td>0.052208</td>\n",
       "      <td>-0.574719</td>\n",
       "      <td>0.218682</td>\n",
       "      <td>-0.074750</td>\n",
       "      <td>-0.551760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>0.280082</td>\n",
       "      <td>0.088898</td>\n",
       "      <td>0.032321</td>\n",
       "      <td>0.446598</td>\n",
       "      <td>-0.037188</td>\n",
       "      <td>-0.228207</td>\n",
       "      <td>0.110942</td>\n",
       "      <td>-0.406541</td>\n",
       "      <td>0.153979</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385241</td>\n",
       "      <td>-0.476314</td>\n",
       "      <td>-0.000984</td>\n",
       "      <td>0.013225</td>\n",
       "      <td>0.294002</td>\n",
       "      <td>0.137939</td>\n",
       "      <td>-0.245951</td>\n",
       "      <td>0.111173</td>\n",
       "      <td>-0.227417</td>\n",
       "      <td>-0.349564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    DepMap_ID  A1BG (1)  A1CF (29974)   A2M (2)  A2ML1 (144568)   \n",
       "0  ACH-000004  0.181332      0.089101 -0.193867       -0.024587  \\\n",
       "1  ACH-000005 -0.076383      0.245190  0.191238        0.153008   \n",
       "2  ACH-000007  0.102195      0.092449 -0.045926        0.171892   \n",
       "3  ACH-000009  0.142342     -0.033126 -0.051224        0.060560   \n",
       "4  ACH-000011  0.280082      0.088898  0.032321        0.446598   \n",
       "\n",
       "   A3GALT2 (127550)  A4GALT (53947)  A4GNT (51146)  AAAS (8086)  AACS (65985)   \n",
       "0          0.038458       -0.181824       0.351672    -0.440073      0.292582  \\\n",
       "1         -0.197035       -0.323295       0.252522    -0.571498     -0.068945   \n",
       "2          0.140561        0.170971       0.086060    -0.432320      0.010664   \n",
       "3          0.116002       -0.010044       0.104725    -0.610481      0.181508   \n",
       "4         -0.037188       -0.228207       0.110942    -0.406541      0.153979   \n",
       "\n",
       "   ...  ZWILCH (55055)  ZWINT (11130)  ZXDA (7789)  ZXDB (158586)   \n",
       "0  ...       -0.124218      -0.469306          NaN            NaN  \\\n",
       "1  ...       -0.212442      -0.426151    -0.068295       0.031635   \n",
       "2  ...       -0.083183      -0.269196     0.101146       0.277820   \n",
       "3  ...       -0.277264      -0.307018     0.044741       0.201551   \n",
       "4  ...       -0.385241      -0.476314    -0.000984       0.013225   \n",
       "\n",
       "   ZXDC (79364)  ZYG11A (440590)  ZYG11B (79699)  ZYX (7791)  ZZEF1 (23140)   \n",
       "0      0.257361         0.244834       -0.408512    0.284734       0.226412  \\\n",
       "1      0.205469        -0.068536       -0.092858    0.074640       0.028779   \n",
       "2      0.208814         0.001393       -0.327514    0.048714      -0.372854   \n",
       "3      0.083866         0.052208       -0.574719    0.218682      -0.074750   \n",
       "4      0.294002         0.137939       -0.245951    0.111173      -0.227417   \n",
       "\n",
       "   ZZZ3 (26009)  \n",
       "0     -0.149566  \n",
       "1     -0.264840  \n",
       "2     -0.433157  \n",
       "3     -0.551760  \n",
       "4     -0.349564  \n",
       "\n",
       "[5 rows x 18120 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_effect = pd.read_csv(\"20Q4v2_Achilles_gene_effect.csv\")\n",
    "\n",
    "df_gene_effect.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DICER1 correlation data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_cor</th>\n",
       "      <th>DICER1_p</th>\n",
       "      <th>gene_selected</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>avg_pos</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>rank</th>\n",
       "      <th>is_on_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DICER1</td>\n",
       "      <td>1</td>\n",
       "      <td>Inf</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>95119894,75</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XPO5</td>\n",
       "      <td>0,503616844</td>\n",
       "      <td>52,45448018</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>43568439,25</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DROSHA</td>\n",
       "      <td>0,33792351</td>\n",
       "      <td>22,30645548</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>31526452,5</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGO2</td>\n",
       "      <td>0,329596473</td>\n",
       "      <td>21,20102211</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>140559591,8</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANKRD52</td>\n",
       "      <td>0,298301131</td>\n",
       "      <td>17,34229262</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>56255045,25</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene     gene_cor     DICER1_p gene_selected genome_alignment   \n",
       "0   DICER1            1          Inf        DICER1            chr14  \\\n",
       "1     XPO5  0,503616844  52,45448018        DICER1             chr6   \n",
       "2   DROSHA   0,33792351  22,30645548        DICER1             chr5   \n",
       "3     AGO2  0,329596473  21,20102211        DICER1             chr8   \n",
       "4  ANKRD52  0,298301131  17,34229262        DICER1            chr12   \n",
       "\n",
       "       avg_pos  chromosome_name  rank  is_on_pathway  \n",
       "0  95119894,75               14     1              1  \n",
       "1  43568439,25                6     2              1  \n",
       "2   31526452,5                5     3              1  \n",
       "3  140559591,8                8     4              1  \n",
       "4  56255045,25               12     5              1  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_correlations = pd.read_csv(\"correlations_DICER1_by_chr.csv\", delimiter=\";\")\n",
    "\n",
    "df_correlations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(('DICER1',\n",
       "  'XPO5',\n",
       "  'DROSHA',\n",
       "  'AGO2',\n",
       "  'ANKRD52',\n",
       "  'DGCR8',\n",
       "  'DOLK',\n",
       "  'MNAT1',\n",
       "  'MAP4K4',\n",
       "  'RBSN',\n",
       "  'RHAG',\n",
       "  'STAT6',\n",
       "  'JAG1',\n",
       "  'HIRIP3',\n",
       "  'NAT1',\n",
       "  'ZSCAN22',\n",
       "  'PCDHB16',\n",
       "  'ORMDL2',\n",
       "  'HYOU1',\n",
       "  'BTG1'),\n",
       " 20)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: Consider making gene selection even easier?\n",
    "\n",
    "number_of_genes = 20 # Should be an even number\n",
    "assert(number_of_genes % 2 == 0)\n",
    "selected_columns = pd.concat([df_correlations[\"gene\"][:number_of_genes // 2], df_correlations[\"gene\"][-number_of_genes // 2:]]).to_list()\n",
    "selected_columns = tuple(selected_columns)\n",
    "selected_columns, len(selected_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['AGO2 (27161)',\n",
       "  'ANKRD52 (283373)',\n",
       "  'BTG1 (694)',\n",
       "  'DGCR8 (54487)',\n",
       "  'DICER1 (23405)',\n",
       "  'DOLK (22845)',\n",
       "  'DROSHA (29102)',\n",
       "  'HIRIP3 (8479)',\n",
       "  'HYOU1 (10525)',\n",
       "  'JAG1 (182)',\n",
       "  'MAP4K4 (9448)',\n",
       "  'MNAT1 (4331)',\n",
       "  'NAT1 (9)',\n",
       "  'ORMDL2 (29095)',\n",
       "  'PCDHB16 (57717)',\n",
       "  'RBSN (64145)',\n",
       "  'RHAG (6005)',\n",
       "  'STAT6 (6778)',\n",
       "  'XPO5 (57510)',\n",
       "  'ZSCAN22 (342945)'],\n",
       " 20)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns_of_interest = [column for column in df_gene_effect.columns if column.split(\" \")[0] in selected_columns]\n",
    "columns_of_interest, len(columns_of_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(selected_columns) == len(columns_of_interest)) # Rough automatic check for whether we found all corresponding columns in gene_effect dataset that we've selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGO2 (27161)</th>\n",
       "      <th>ANKRD52 (283373)</th>\n",
       "      <th>BTG1 (694)</th>\n",
       "      <th>DGCR8 (54487)</th>\n",
       "      <th>DICER1 (23405)</th>\n",
       "      <th>DOLK (22845)</th>\n",
       "      <th>DROSHA (29102)</th>\n",
       "      <th>HIRIP3 (8479)</th>\n",
       "      <th>HYOU1 (10525)</th>\n",
       "      <th>JAG1 (182)</th>\n",
       "      <th>MAP4K4 (9448)</th>\n",
       "      <th>MNAT1 (4331)</th>\n",
       "      <th>NAT1 (9)</th>\n",
       "      <th>ORMDL2 (29095)</th>\n",
       "      <th>PCDHB16 (57717)</th>\n",
       "      <th>RBSN (64145)</th>\n",
       "      <th>RHAG (6005)</th>\n",
       "      <th>STAT6 (6778)</th>\n",
       "      <th>XPO5 (57510)</th>\n",
       "      <th>ZSCAN22 (342945)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.485791</td>\n",
       "      <td>0.061758</td>\n",
       "      <td>0.009099</td>\n",
       "      <td>-0.412221</td>\n",
       "      <td>-0.800483</td>\n",
       "      <td>-0.982393</td>\n",
       "      <td>-0.471354</td>\n",
       "      <td>-0.044968</td>\n",
       "      <td>-1.248355</td>\n",
       "      <td>-0.286972</td>\n",
       "      <td>-0.017386</td>\n",
       "      <td>-1.209837</td>\n",
       "      <td>0.163402</td>\n",
       "      <td>0.152011</td>\n",
       "      <td>-0.005288</td>\n",
       "      <td>-0.233983</td>\n",
       "      <td>-0.022268</td>\n",
       "      <td>-0.263227</td>\n",
       "      <td>-0.976411</td>\n",
       "      <td>-0.004016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.211648</td>\n",
       "      <td>0.144988</td>\n",
       "      <td>-0.279494</td>\n",
       "      <td>-0.711791</td>\n",
       "      <td>-1.342802</td>\n",
       "      <td>-1.150483</td>\n",
       "      <td>-0.779502</td>\n",
       "      <td>-0.000837</td>\n",
       "      <td>-1.166890</td>\n",
       "      <td>-0.060406</td>\n",
       "      <td>-0.130849</td>\n",
       "      <td>-1.125206</td>\n",
       "      <td>0.313035</td>\n",
       "      <td>-0.055905</td>\n",
       "      <td>-0.257466</td>\n",
       "      <td>-0.569009</td>\n",
       "      <td>-0.162031</td>\n",
       "      <td>0.168497</td>\n",
       "      <td>-0.859519</td>\n",
       "      <td>-0.028137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.047294</td>\n",
       "      <td>-0.035615</td>\n",
       "      <td>0.044191</td>\n",
       "      <td>-0.409023</td>\n",
       "      <td>-0.993394</td>\n",
       "      <td>-0.843561</td>\n",
       "      <td>-0.261872</td>\n",
       "      <td>-0.021099</td>\n",
       "      <td>-1.179217</td>\n",
       "      <td>-0.029307</td>\n",
       "      <td>0.258903</td>\n",
       "      <td>-1.402721</td>\n",
       "      <td>0.335307</td>\n",
       "      <td>0.074292</td>\n",
       "      <td>-0.400558</td>\n",
       "      <td>0.177025</td>\n",
       "      <td>0.115254</td>\n",
       "      <td>-0.027024</td>\n",
       "      <td>-0.910107</td>\n",
       "      <td>0.063903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.147972</td>\n",
       "      <td>-0.055845</td>\n",
       "      <td>0.110307</td>\n",
       "      <td>-0.349032</td>\n",
       "      <td>-0.409258</td>\n",
       "      <td>-0.987303</td>\n",
       "      <td>-0.302471</td>\n",
       "      <td>0.108021</td>\n",
       "      <td>-1.397957</td>\n",
       "      <td>0.239571</td>\n",
       "      <td>0.051585</td>\n",
       "      <td>-1.187064</td>\n",
       "      <td>0.119674</td>\n",
       "      <td>0.016975</td>\n",
       "      <td>-0.112180</td>\n",
       "      <td>-0.191086</td>\n",
       "      <td>0.246043</td>\n",
       "      <td>-0.098910</td>\n",
       "      <td>-0.834269</td>\n",
       "      <td>-0.036320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.244794</td>\n",
       "      <td>-0.083400</td>\n",
       "      <td>-0.122937</td>\n",
       "      <td>-0.434856</td>\n",
       "      <td>-0.792070</td>\n",
       "      <td>-0.933021</td>\n",
       "      <td>-0.407159</td>\n",
       "      <td>-0.128025</td>\n",
       "      <td>-1.400010</td>\n",
       "      <td>-0.128237</td>\n",
       "      <td>0.175264</td>\n",
       "      <td>-1.455323</td>\n",
       "      <td>0.069390</td>\n",
       "      <td>0.175852</td>\n",
       "      <td>0.109310</td>\n",
       "      <td>-0.558395</td>\n",
       "      <td>0.046969</td>\n",
       "      <td>-0.022238</td>\n",
       "      <td>-0.823367</td>\n",
       "      <td>0.113082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>-0.500938</td>\n",
       "      <td>-0.567413</td>\n",
       "      <td>-0.103990</td>\n",
       "      <td>-0.472353</td>\n",
       "      <td>-0.959784</td>\n",
       "      <td>-0.513790</td>\n",
       "      <td>-0.750916</td>\n",
       "      <td>-0.010317</td>\n",
       "      <td>-1.452658</td>\n",
       "      <td>0.017472</td>\n",
       "      <td>0.746325</td>\n",
       "      <td>-0.892737</td>\n",
       "      <td>-0.028297</td>\n",
       "      <td>0.046277</td>\n",
       "      <td>-0.066448</td>\n",
       "      <td>-0.776029</td>\n",
       "      <td>0.171691</td>\n",
       "      <td>-0.180649</td>\n",
       "      <td>-1.148982</td>\n",
       "      <td>0.075566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>-0.507409</td>\n",
       "      <td>-0.505528</td>\n",
       "      <td>-0.172067</td>\n",
       "      <td>-0.836671</td>\n",
       "      <td>-1.603509</td>\n",
       "      <td>-0.534013</td>\n",
       "      <td>-1.300670</td>\n",
       "      <td>-0.207670</td>\n",
       "      <td>-1.373992</td>\n",
       "      <td>0.347200</td>\n",
       "      <td>0.712036</td>\n",
       "      <td>-0.963081</td>\n",
       "      <td>0.045723</td>\n",
       "      <td>0.075520</td>\n",
       "      <td>-0.241542</td>\n",
       "      <td>-0.912934</td>\n",
       "      <td>0.161383</td>\n",
       "      <td>-0.126069</td>\n",
       "      <td>-1.631937</td>\n",
       "      <td>0.096254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805</th>\n",
       "      <td>-0.303557</td>\n",
       "      <td>-0.226048</td>\n",
       "      <td>-0.145903</td>\n",
       "      <td>-0.512966</td>\n",
       "      <td>-0.777286</td>\n",
       "      <td>-0.736989</td>\n",
       "      <td>-0.835479</td>\n",
       "      <td>-0.127032</td>\n",
       "      <td>-1.444140</td>\n",
       "      <td>-0.118616</td>\n",
       "      <td>0.713992</td>\n",
       "      <td>-0.950609</td>\n",
       "      <td>0.069828</td>\n",
       "      <td>-0.068681</td>\n",
       "      <td>-0.098058</td>\n",
       "      <td>-0.718332</td>\n",
       "      <td>0.085602</td>\n",
       "      <td>-0.235009</td>\n",
       "      <td>-0.980329</td>\n",
       "      <td>-0.149162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>-0.208096</td>\n",
       "      <td>-0.440253</td>\n",
       "      <td>-0.267566</td>\n",
       "      <td>-0.336869</td>\n",
       "      <td>-1.030868</td>\n",
       "      <td>-0.600950</td>\n",
       "      <td>-0.883779</td>\n",
       "      <td>-0.000932</td>\n",
       "      <td>-1.470970</td>\n",
       "      <td>0.069590</td>\n",
       "      <td>0.775574</td>\n",
       "      <td>-0.879057</td>\n",
       "      <td>0.137400</td>\n",
       "      <td>-0.007353</td>\n",
       "      <td>-0.063559</td>\n",
       "      <td>-0.092920</td>\n",
       "      <td>0.170532</td>\n",
       "      <td>-0.159439</td>\n",
       "      <td>-0.686586</td>\n",
       "      <td>0.044028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>-0.235582</td>\n",
       "      <td>-0.419662</td>\n",
       "      <td>0.120573</td>\n",
       "      <td>-0.400833</td>\n",
       "      <td>-1.120057</td>\n",
       "      <td>-0.981647</td>\n",
       "      <td>-0.800313</td>\n",
       "      <td>0.178577</td>\n",
       "      <td>-1.544684</td>\n",
       "      <td>-0.004937</td>\n",
       "      <td>0.375205</td>\n",
       "      <td>-1.220549</td>\n",
       "      <td>-0.011369</td>\n",
       "      <td>0.108498</td>\n",
       "      <td>-0.136718</td>\n",
       "      <td>-0.705100</td>\n",
       "      <td>0.199102</td>\n",
       "      <td>-0.300074</td>\n",
       "      <td>-1.092880</td>\n",
       "      <td>0.100196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     AGO2 (27161)  ANKRD52 (283373)  BTG1 (694)  DGCR8 (54487)   \n",
       "0       -0.485791          0.061758    0.009099      -0.412221  \\\n",
       "1       -0.211648          0.144988   -0.279494      -0.711791   \n",
       "2       -0.047294         -0.035615    0.044191      -0.409023   \n",
       "3        0.147972         -0.055845    0.110307      -0.349032   \n",
       "4       -0.244794         -0.083400   -0.122937      -0.434856   \n",
       "..            ...               ...         ...            ...   \n",
       "803     -0.500938         -0.567413   -0.103990      -0.472353   \n",
       "804     -0.507409         -0.505528   -0.172067      -0.836671   \n",
       "805     -0.303557         -0.226048   -0.145903      -0.512966   \n",
       "806     -0.208096         -0.440253   -0.267566      -0.336869   \n",
       "807     -0.235582         -0.419662    0.120573      -0.400833   \n",
       "\n",
       "     DICER1 (23405)  DOLK (22845)  DROSHA (29102)  HIRIP3 (8479)   \n",
       "0         -0.800483     -0.982393       -0.471354      -0.044968  \\\n",
       "1         -1.342802     -1.150483       -0.779502      -0.000837   \n",
       "2         -0.993394     -0.843561       -0.261872      -0.021099   \n",
       "3         -0.409258     -0.987303       -0.302471       0.108021   \n",
       "4         -0.792070     -0.933021       -0.407159      -0.128025   \n",
       "..              ...           ...             ...            ...   \n",
       "803       -0.959784     -0.513790       -0.750916      -0.010317   \n",
       "804       -1.603509     -0.534013       -1.300670      -0.207670   \n",
       "805       -0.777286     -0.736989       -0.835479      -0.127032   \n",
       "806       -1.030868     -0.600950       -0.883779      -0.000932   \n",
       "807       -1.120057     -0.981647       -0.800313       0.178577   \n",
       "\n",
       "     HYOU1 (10525)  JAG1 (182)  MAP4K4 (9448)  MNAT1 (4331)  NAT1 (9)   \n",
       "0        -1.248355   -0.286972      -0.017386     -1.209837  0.163402  \\\n",
       "1        -1.166890   -0.060406      -0.130849     -1.125206  0.313035   \n",
       "2        -1.179217   -0.029307       0.258903     -1.402721  0.335307   \n",
       "3        -1.397957    0.239571       0.051585     -1.187064  0.119674   \n",
       "4        -1.400010   -0.128237       0.175264     -1.455323  0.069390   \n",
       "..             ...         ...            ...           ...       ...   \n",
       "803      -1.452658    0.017472       0.746325     -0.892737 -0.028297   \n",
       "804      -1.373992    0.347200       0.712036     -0.963081  0.045723   \n",
       "805      -1.444140   -0.118616       0.713992     -0.950609  0.069828   \n",
       "806      -1.470970    0.069590       0.775574     -0.879057  0.137400   \n",
       "807      -1.544684   -0.004937       0.375205     -1.220549 -0.011369   \n",
       "\n",
       "     ORMDL2 (29095)  PCDHB16 (57717)  RBSN (64145)  RHAG (6005)  STAT6 (6778)   \n",
       "0          0.152011        -0.005288     -0.233983    -0.022268     -0.263227  \\\n",
       "1         -0.055905        -0.257466     -0.569009    -0.162031      0.168497   \n",
       "2          0.074292        -0.400558      0.177025     0.115254     -0.027024   \n",
       "3          0.016975        -0.112180     -0.191086     0.246043     -0.098910   \n",
       "4          0.175852         0.109310     -0.558395     0.046969     -0.022238   \n",
       "..              ...              ...           ...          ...           ...   \n",
       "803        0.046277        -0.066448     -0.776029     0.171691     -0.180649   \n",
       "804        0.075520        -0.241542     -0.912934     0.161383     -0.126069   \n",
       "805       -0.068681        -0.098058     -0.718332     0.085602     -0.235009   \n",
       "806       -0.007353        -0.063559     -0.092920     0.170532     -0.159439   \n",
       "807        0.108498        -0.136718     -0.705100     0.199102     -0.300074   \n",
       "\n",
       "     XPO5 (57510)  ZSCAN22 (342945)  \n",
       "0       -0.976411         -0.004016  \n",
       "1       -0.859519         -0.028137  \n",
       "2       -0.910107          0.063903  \n",
       "3       -0.834269         -0.036320  \n",
       "4       -0.823367          0.113082  \n",
       "..            ...               ...  \n",
       "803     -1.148982          0.075566  \n",
       "804     -1.631937          0.096254  \n",
       "805     -0.980329         -0.149162  \n",
       "806     -0.686586          0.044028  \n",
       "807     -1.092880          0.100196  \n",
       "\n",
       "[808 rows x 20 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_effect[columns_of_interest]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(796, 20)"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = df_gene_effect[columns_of_interest].dropna() # Is dropping rows with na a good idea? Maybe use mean value instead?\n",
    "X = X.to_numpy()\n",
    "X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "# Get the current working directory (where the notebook is running)\n",
    "notebook_dir = os.getcwd()\n",
    "\n",
    "# Go up one level to the parent directory\n",
    "parent_dir = os.path.abspath(os.path.join(notebook_dir, '..'))\n",
    "\n",
    "# Add parent directory to sys.path\n",
    "if parent_dir not in sys.path:\n",
    "    sys.path.insert(0, parent_dir)\n",
    "\n",
    "# Now you can import your module\n",
    "from utils import factor_model_solution, factor_goodness_of_fit_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.46006248],\n",
       "       [-0.43950578],\n",
       "       [ 0.0853125 ],\n",
       "       [-0.32229552],\n",
       "       [-0.78339134],\n",
       "       [-0.28554635],\n",
       "       [-0.4212857 ],\n",
       "       [ 0.13032313],\n",
       "       [-0.14649522],\n",
       "       [-0.1262754 ],\n",
       "       [ 0.19442815],\n",
       "       [-0.23221741],\n",
       "       [ 0.12352559],\n",
       "       [ 0.13787435],\n",
       "       [-0.08899356],\n",
       "       [-0.29076869],\n",
       "       [-0.14821865],\n",
       "       [ 0.10028168],\n",
       "       [-0.64101898],\n",
       "       [-0.09776291]])"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, lambda_hat = factor_model_solution(X, k = 1)\n",
    "lambda_hat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>is_on_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DICER1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XPO5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DROSHA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGO2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANKRD52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DGCR8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DOLK</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MNAT1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MAP4K4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RBSN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>RHAG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>STAT6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>JAG1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>HIRIP3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>NAT1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>ZSCAN22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>PCDHB16</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>ORMDL2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>HYOU1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>BTG1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  is_on_pathway\n",
       "0     DICER1              1\n",
       "1       XPO5              1\n",
       "2     DROSHA              1\n",
       "3       AGO2              1\n",
       "4    ANKRD52              1\n",
       "5      DGCR8              1\n",
       "6       DOLK              0\n",
       "7      MNAT1              0\n",
       "8     MAP4K4              0\n",
       "9       RBSN              0\n",
       "490     RHAG              0\n",
       "491    STAT6              0\n",
       "492     JAG1              0\n",
       "493   HIRIP3              0\n",
       "494     NAT1              0\n",
       "495  ZSCAN22              0\n",
       "496  PCDHB16              0\n",
       "497   ORMDL2              0\n",
       "498    HYOU1              0\n",
       "499     BTG1              0"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([df_correlations[[\"gene\", \"is_on_pathway\"]][:number_of_genes // 2], df_correlations[[\"gene\", \"is_on_pathway\"]][-number_of_genes // 2:]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AGO2 (27161)',\n",
       " 'ANKRD52 (283373)',\n",
       " 'BTG1 (694)',\n",
       " 'DGCR8 (54487)',\n",
       " 'DICER1 (23405)',\n",
       " 'DOLK (22845)',\n",
       " 'DROSHA (29102)',\n",
       " 'HIRIP3 (8479)',\n",
       " 'HYOU1 (10525)',\n",
       " 'JAG1 (182)',\n",
       " 'MAP4K4 (9448)',\n",
       " 'MNAT1 (4331)',\n",
       " 'NAT1 (9)',\n",
       " 'ORMDL2 (29095)',\n",
       " 'PCDHB16 (57717)',\n",
       " 'RBSN (64145)',\n",
       " 'RHAG (6005)',\n",
       " 'STAT6 (6778)',\n",
       " 'XPO5 (57510)',\n",
       " 'ZSCAN22 (342945)']"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns_of_interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       gene  loadings  is_on_pathway\n",
      "10   MAP4K4  0.194428              0\n",
      "13   ORMDL2  0.137874              0\n",
      "7    HIRIP3  0.130323              0\n",
      "12     NAT1  0.123526              0\n",
      "17    STAT6  0.100282              0\n",
      "2      BTG1  0.085313              0\n",
      "14  PCDHB16 -0.088994              0\n",
      "19  ZSCAN22 -0.097763              0\n",
      "9      JAG1 -0.126275              0\n",
      "8     HYOU1 -0.146495              0\n",
      "16     RHAG -0.148219              0\n",
      "11    MNAT1 -0.232217              0\n",
      "5      DOLK -0.285546              0\n",
      "15     RBSN -0.290769              0\n",
      "3     DGCR8 -0.322296              1\n",
      "6    DROSHA -0.421286              1\n",
      "1   ANKRD52 -0.439506              1\n",
      "0      AGO2 -0.460062              1\n",
      "18     XPO5 -0.641019              1\n",
      "4    DICER1 -0.783391              1\n"
     ]
    }
   ],
   "source": [
    "loadings = lambda_hat.T[0] # Extract loadings as a list\n",
    "\n",
    "# Extract gene names (everything before the space)\n",
    "genes = [column.split(\" \")[0] for column in columns_of_interest]\n",
    "\n",
    "# Create a DataFrame from genes and values\n",
    "df_values = pd.DataFrame({\n",
    "    'gene': genes,\n",
    "    'loadings': loadings\n",
    "})\n",
    "\n",
    "# Merge with df_correlations on 'gene'\n",
    "merged_df = df_values.merge(df_correlations[['gene', 'is_on_pathway']], on='gene', how='left')\n",
    "\n",
    "print(merged_df.sort_values(\"loadings\", ascending=False))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We get heavier loadings on genes that actually lie on the pathway!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.191027767126529e-31"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "factor_goodness_of_fit_test(X, k = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test shows that k = 1 **is not** sufficient to describe (**all**) the data. But sufficient to **only** capture the (single) pathway perhaps?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
