{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing the data for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>A1BG (1)</th>\n",
       "      <th>A1CF (29974)</th>\n",
       "      <th>A2M (2)</th>\n",
       "      <th>A2ML1 (144568)</th>\n",
       "      <th>A3GALT2 (127550)</th>\n",
       "      <th>A4GALT (53947)</th>\n",
       "      <th>A4GNT (51146)</th>\n",
       "      <th>AAAS (8086)</th>\n",
       "      <th>AACS (65985)</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH (55055)</th>\n",
       "      <th>ZWINT (11130)</th>\n",
       "      <th>ZXDA (7789)</th>\n",
       "      <th>ZXDB (158586)</th>\n",
       "      <th>ZXDC (79364)</th>\n",
       "      <th>ZYG11A (440590)</th>\n",
       "      <th>ZYG11B (79699)</th>\n",
       "      <th>ZYX (7791)</th>\n",
       "      <th>ZZEF1 (23140)</th>\n",
       "      <th>ZZZ3 (26009)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>0.181332</td>\n",
       "      <td>0.089101</td>\n",
       "      <td>-0.193867</td>\n",
       "      <td>-0.024587</td>\n",
       "      <td>0.038458</td>\n",
       "      <td>-0.181824</td>\n",
       "      <td>0.351672</td>\n",
       "      <td>-0.440073</td>\n",
       "      <td>0.292582</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.124218</td>\n",
       "      <td>-0.469306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.257361</td>\n",
       "      <td>0.244834</td>\n",
       "      <td>-0.408512</td>\n",
       "      <td>0.284734</td>\n",
       "      <td>0.226412</td>\n",
       "      <td>-0.149566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>-0.076383</td>\n",
       "      <td>0.245190</td>\n",
       "      <td>0.191238</td>\n",
       "      <td>0.153008</td>\n",
       "      <td>-0.197035</td>\n",
       "      <td>-0.323295</td>\n",
       "      <td>0.252522</td>\n",
       "      <td>-0.571498</td>\n",
       "      <td>-0.068945</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.212442</td>\n",
       "      <td>-0.426151</td>\n",
       "      <td>-0.068295</td>\n",
       "      <td>0.031635</td>\n",
       "      <td>0.205469</td>\n",
       "      <td>-0.068536</td>\n",
       "      <td>-0.092858</td>\n",
       "      <td>0.074640</td>\n",
       "      <td>0.028779</td>\n",
       "      <td>-0.264840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>0.102195</td>\n",
       "      <td>0.092449</td>\n",
       "      <td>-0.045926</td>\n",
       "      <td>0.171892</td>\n",
       "      <td>0.140561</td>\n",
       "      <td>0.170971</td>\n",
       "      <td>0.086060</td>\n",
       "      <td>-0.432320</td>\n",
       "      <td>0.010664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.083183</td>\n",
       "      <td>-0.269196</td>\n",
       "      <td>0.101146</td>\n",
       "      <td>0.277820</td>\n",
       "      <td>0.208814</td>\n",
       "      <td>0.001393</td>\n",
       "      <td>-0.327514</td>\n",
       "      <td>0.048714</td>\n",
       "      <td>-0.372854</td>\n",
       "      <td>-0.433157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000009</td>\n",
       "      <td>0.142342</td>\n",
       "      <td>-0.033126</td>\n",
       "      <td>-0.051224</td>\n",
       "      <td>0.060560</td>\n",
       "      <td>0.116002</td>\n",
       "      <td>-0.010044</td>\n",
       "      <td>0.104725</td>\n",
       "      <td>-0.610481</td>\n",
       "      <td>0.181508</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.277264</td>\n",
       "      <td>-0.307018</td>\n",
       "      <td>0.044741</td>\n",
       "      <td>0.201551</td>\n",
       "      <td>0.083866</td>\n",
       "      <td>0.052208</td>\n",
       "      <td>-0.574719</td>\n",
       "      <td>0.218682</td>\n",
       "      <td>-0.074750</td>\n",
       "      <td>-0.551760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>0.280082</td>\n",
       "      <td>0.088898</td>\n",
       "      <td>0.032321</td>\n",
       "      <td>0.446598</td>\n",
       "      <td>-0.037188</td>\n",
       "      <td>-0.228207</td>\n",
       "      <td>0.110942</td>\n",
       "      <td>-0.406541</td>\n",
       "      <td>0.153979</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385241</td>\n",
       "      <td>-0.476314</td>\n",
       "      <td>-0.000984</td>\n",
       "      <td>0.013225</td>\n",
       "      <td>0.294002</td>\n",
       "      <td>0.137939</td>\n",
       "      <td>-0.245951</td>\n",
       "      <td>0.111173</td>\n",
       "      <td>-0.227417</td>\n",
       "      <td>-0.349564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    DepMap_ID  A1BG (1)  A1CF (29974)   A2M (2)  A2ML1 (144568)   \n",
       "0  ACH-000004  0.181332      0.089101 -0.193867       -0.024587  \\\n",
       "1  ACH-000005 -0.076383      0.245190  0.191238        0.153008   \n",
       "2  ACH-000007  0.102195      0.092449 -0.045926        0.171892   \n",
       "3  ACH-000009  0.142342     -0.033126 -0.051224        0.060560   \n",
       "4  ACH-000011  0.280082      0.088898  0.032321        0.446598   \n",
       "\n",
       "   A3GALT2 (127550)  A4GALT (53947)  A4GNT (51146)  AAAS (8086)  AACS (65985)   \n",
       "0          0.038458       -0.181824       0.351672    -0.440073      0.292582  \\\n",
       "1         -0.197035       -0.323295       0.252522    -0.571498     -0.068945   \n",
       "2          0.140561        0.170971       0.086060    -0.432320      0.010664   \n",
       "3          0.116002       -0.010044       0.104725    -0.610481      0.181508   \n",
       "4         -0.037188       -0.228207       0.110942    -0.406541      0.153979   \n",
       "\n",
       "   ...  ZWILCH (55055)  ZWINT (11130)  ZXDA (7789)  ZXDB (158586)   \n",
       "0  ...       -0.124218      -0.469306          NaN            NaN  \\\n",
       "1  ...       -0.212442      -0.426151    -0.068295       0.031635   \n",
       "2  ...       -0.083183      -0.269196     0.101146       0.277820   \n",
       "3  ...       -0.277264      -0.307018     0.044741       0.201551   \n",
       "4  ...       -0.385241      -0.476314    -0.000984       0.013225   \n",
       "\n",
       "   ZXDC (79364)  ZYG11A (440590)  ZYG11B (79699)  ZYX (7791)  ZZEF1 (23140)   \n",
       "0      0.257361         0.244834       -0.408512    0.284734       0.226412  \\\n",
       "1      0.205469        -0.068536       -0.092858    0.074640       0.028779   \n",
       "2      0.208814         0.001393       -0.327514    0.048714      -0.372854   \n",
       "3      0.083866         0.052208       -0.574719    0.218682      -0.074750   \n",
       "4      0.294002         0.137939       -0.245951    0.111173      -0.227417   \n",
       "\n",
       "   ZZZ3 (26009)  \n",
       "0     -0.149566  \n",
       "1     -0.264840  \n",
       "2     -0.433157  \n",
       "3     -0.551760  \n",
       "4     -0.349564  \n",
       "\n",
       "[5 rows x 18120 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_effect = pd.read_csv(\"20Q4v2_Achilles_gene_effect.csv\")\n",
    "\n",
    "df_gene_effect.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_cor</th>\n",
       "      <th>DICER1_p</th>\n",
       "      <th>gene_selected</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>avg_pos</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>rank</th>\n",
       "      <th>is_on_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DICER1</td>\n",
       "      <td>1</td>\n",
       "      <td>Inf</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>95119894,75</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XPO5</td>\n",
       "      <td>0,503616844</td>\n",
       "      <td>52,45448018</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>43568439,25</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DROSHA</td>\n",
       "      <td>0,33792351</td>\n",
       "      <td>22,30645548</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>31526452,5</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGO2</td>\n",
       "      <td>0,329596473</td>\n",
       "      <td>21,20102211</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>140559591,8</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANKRD52</td>\n",
       "      <td>0,298301131</td>\n",
       "      <td>17,34229262</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>56255045,25</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene     gene_cor     DICER1_p gene_selected genome_alignment   \n",
       "0   DICER1            1          Inf        DICER1            chr14  \\\n",
       "1     XPO5  0,503616844  52,45448018        DICER1             chr6   \n",
       "2   DROSHA   0,33792351  22,30645548        DICER1             chr5   \n",
       "3     AGO2  0,329596473  21,20102211        DICER1             chr8   \n",
       "4  ANKRD52  0,298301131  17,34229262        DICER1            chr12   \n",
       "\n",
       "       avg_pos  chromosome_name  rank  is_on_pathway  \n",
       "0  95119894,75               14     1              1  \n",
       "1  43568439,25                6     2              1  \n",
       "2   31526452,5                5     3              1  \n",
       "3  140559591,8                8     4              1  \n",
       "4  56255045,25               12     5              1  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_correlations = pd.read_csv(\"correlations_DICER1_by_chr.csv\", delimiter=\";\")\n",
    "\n",
    "df_correlations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(('DICER1',\n",
       "  'XPO5',\n",
       "  'DROSHA',\n",
       "  'AGO2',\n",
       "  'ANKRD52',\n",
       "  'ZSCAN22',\n",
       "  'PCDHB16',\n",
       "  'ORMDL2',\n",
       "  'HYOU1',\n",
       "  'BTG1'),\n",
       " 10)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: Make gene selection easier\n",
    "selected_columns = pd.concat([df_correlations[\"gene\"][:5], df_correlations[\"gene\"][-5:]]).to_list()\n",
    "selected_columns = tuple(selected_columns)\n",
    "selected_columns, len(selected_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['AGO2 (27161)',\n",
       "  'ANKRD52 (283373)',\n",
       "  'BTG1 (694)',\n",
       "  'DICER1 (23405)',\n",
       "  'DROSHA (29102)',\n",
       "  'HYOU1 (10525)',\n",
       "  'ORMDL2 (29095)',\n",
       "  'PCDHB16 (57717)',\n",
       "  'XPO5 (57510)',\n",
       "  'ZSCAN22 (342945)'],\n",
       " 10)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns_of_interest = [column for column in df_gene_effect.columns if column.startswith(selected_columns)]\n",
    "columns_of_interest, len(columns_of_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(selected_columns) == len(columns_of_interest)) # Rough automatic check for whether we found all corresponding columns in gene_effect dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGO2 (27161)</th>\n",
       "      <th>ANKRD52 (283373)</th>\n",
       "      <th>BTG1 (694)</th>\n",
       "      <th>DICER1 (23405)</th>\n",
       "      <th>DROSHA (29102)</th>\n",
       "      <th>HYOU1 (10525)</th>\n",
       "      <th>ORMDL2 (29095)</th>\n",
       "      <th>PCDHB16 (57717)</th>\n",
       "      <th>XPO5 (57510)</th>\n",
       "      <th>ZSCAN22 (342945)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.485791</td>\n",
       "      <td>0.061758</td>\n",
       "      <td>0.009099</td>\n",
       "      <td>-0.800483</td>\n",
       "      <td>-0.471354</td>\n",
       "      <td>-1.248355</td>\n",
       "      <td>0.152011</td>\n",
       "      <td>-0.005288</td>\n",
       "      <td>-0.976411</td>\n",
       "      <td>-0.004016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.211648</td>\n",
       "      <td>0.144988</td>\n",
       "      <td>-0.279494</td>\n",
       "      <td>-1.342802</td>\n",
       "      <td>-0.779502</td>\n",
       "      <td>-1.166890</td>\n",
       "      <td>-0.055905</td>\n",
       "      <td>-0.257466</td>\n",
       "      <td>-0.859519</td>\n",
       "      <td>-0.028137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.047294</td>\n",
       "      <td>-0.035615</td>\n",
       "      <td>0.044191</td>\n",
       "      <td>-0.993394</td>\n",
       "      <td>-0.261872</td>\n",
       "      <td>-1.179217</td>\n",
       "      <td>0.074292</td>\n",
       "      <td>-0.400558</td>\n",
       "      <td>-0.910107</td>\n",
       "      <td>0.063903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.147972</td>\n",
       "      <td>-0.055845</td>\n",
       "      <td>0.110307</td>\n",
       "      <td>-0.409258</td>\n",
       "      <td>-0.302471</td>\n",
       "      <td>-1.397957</td>\n",
       "      <td>0.016975</td>\n",
       "      <td>-0.112180</td>\n",
       "      <td>-0.834269</td>\n",
       "      <td>-0.036320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.244794</td>\n",
       "      <td>-0.083400</td>\n",
       "      <td>-0.122937</td>\n",
       "      <td>-0.792070</td>\n",
       "      <td>-0.407159</td>\n",
       "      <td>-1.400010</td>\n",
       "      <td>0.175852</td>\n",
       "      <td>0.109310</td>\n",
       "      <td>-0.823367</td>\n",
       "      <td>0.113082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>-0.500938</td>\n",
       "      <td>-0.567413</td>\n",
       "      <td>-0.103990</td>\n",
       "      <td>-0.959784</td>\n",
       "      <td>-0.750916</td>\n",
       "      <td>-1.452658</td>\n",
       "      <td>0.046277</td>\n",
       "      <td>-0.066448</td>\n",
       "      <td>-1.148982</td>\n",
       "      <td>0.075566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>-0.507409</td>\n",
       "      <td>-0.505528</td>\n",
       "      <td>-0.172067</td>\n",
       "      <td>-1.603509</td>\n",
       "      <td>-1.300670</td>\n",
       "      <td>-1.373992</td>\n",
       "      <td>0.075520</td>\n",
       "      <td>-0.241542</td>\n",
       "      <td>-1.631937</td>\n",
       "      <td>0.096254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805</th>\n",
       "      <td>-0.303557</td>\n",
       "      <td>-0.226048</td>\n",
       "      <td>-0.145903</td>\n",
       "      <td>-0.777286</td>\n",
       "      <td>-0.835479</td>\n",
       "      <td>-1.444140</td>\n",
       "      <td>-0.068681</td>\n",
       "      <td>-0.098058</td>\n",
       "      <td>-0.980329</td>\n",
       "      <td>-0.149162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>-0.208096</td>\n",
       "      <td>-0.440253</td>\n",
       "      <td>-0.267566</td>\n",
       "      <td>-1.030868</td>\n",
       "      <td>-0.883779</td>\n",
       "      <td>-1.470970</td>\n",
       "      <td>-0.007353</td>\n",
       "      <td>-0.063559</td>\n",
       "      <td>-0.686586</td>\n",
       "      <td>0.044028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>-0.235582</td>\n",
       "      <td>-0.419662</td>\n",
       "      <td>0.120573</td>\n",
       "      <td>-1.120057</td>\n",
       "      <td>-0.800313</td>\n",
       "      <td>-1.544684</td>\n",
       "      <td>0.108498</td>\n",
       "      <td>-0.136718</td>\n",
       "      <td>-1.092880</td>\n",
       "      <td>0.100196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     AGO2 (27161)  ANKRD52 (283373)  BTG1 (694)  DICER1 (23405)   \n",
       "0       -0.485791          0.061758    0.009099       -0.800483  \\\n",
       "1       -0.211648          0.144988   -0.279494       -1.342802   \n",
       "2       -0.047294         -0.035615    0.044191       -0.993394   \n",
       "3        0.147972         -0.055845    0.110307       -0.409258   \n",
       "4       -0.244794         -0.083400   -0.122937       -0.792070   \n",
       "..            ...               ...         ...             ...   \n",
       "803     -0.500938         -0.567413   -0.103990       -0.959784   \n",
       "804     -0.507409         -0.505528   -0.172067       -1.603509   \n",
       "805     -0.303557         -0.226048   -0.145903       -0.777286   \n",
       "806     -0.208096         -0.440253   -0.267566       -1.030868   \n",
       "807     -0.235582         -0.419662    0.120573       -1.120057   \n",
       "\n",
       "     DROSHA (29102)  HYOU1 (10525)  ORMDL2 (29095)  PCDHB16 (57717)   \n",
       "0         -0.471354      -1.248355        0.152011        -0.005288  \\\n",
       "1         -0.779502      -1.166890       -0.055905        -0.257466   \n",
       "2         -0.261872      -1.179217        0.074292        -0.400558   \n",
       "3         -0.302471      -1.397957        0.016975        -0.112180   \n",
       "4         -0.407159      -1.400010        0.175852         0.109310   \n",
       "..              ...            ...             ...              ...   \n",
       "803       -0.750916      -1.452658        0.046277        -0.066448   \n",
       "804       -1.300670      -1.373992        0.075520        -0.241542   \n",
       "805       -0.835479      -1.444140       -0.068681        -0.098058   \n",
       "806       -0.883779      -1.470970       -0.007353        -0.063559   \n",
       "807       -0.800313      -1.544684        0.108498        -0.136718   \n",
       "\n",
       "     XPO5 (57510)  ZSCAN22 (342945)  \n",
       "0       -0.976411         -0.004016  \n",
       "1       -0.859519         -0.028137  \n",
       "2       -0.910107          0.063903  \n",
       "3       -0.834269         -0.036320  \n",
       "4       -0.823367          0.113082  \n",
       "..            ...               ...  \n",
       "803     -1.148982          0.075566  \n",
       "804     -1.631937          0.096254  \n",
       "805     -0.980329         -0.149162  \n",
       "806     -0.686586          0.044028  \n",
       "807     -1.092880          0.100196  \n",
       "\n",
       "[808 rows x 10 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_effect[columns_of_interest]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(808, 10)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = df_gene_effect[columns_of_interest].to_numpy()\n",
    "X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "# Get the current working directory (where the notebook is running)\n",
    "notebook_dir = os.getcwd()\n",
    "\n",
    "# Go up one level to the parent directory\n",
    "parent_dir = os.path.abspath(os.path.join(notebook_dir, '..'))\n",
    "\n",
    "# Add parent directory to sys.path\n",
    "if parent_dir not in sys.path:\n",
    "    sys.path.insert(0, parent_dir)\n",
    "\n",
    "# Now you can import your module\n",
    "from utils import factor_model_solution, factor_goodness_of_fit_test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.48939537],\n",
       "       [ 0.48488142],\n",
       "       [-0.09223838],\n",
       "       [ 0.72307398],\n",
       "       [ 0.41419227],\n",
       "       [ 0.12533711],\n",
       "       [-0.13337069],\n",
       "       [ 0.09556584],\n",
       "       [ 0.68065617],\n",
       "       [ 0.07652927]])"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, lambda_hat = factor_model_solution(X, k = 1)\n",
    "lambda_hat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>is_on_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DICER1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XPO5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DROSHA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGO2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANKRD52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>ZSCAN22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>PCDHB16</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>ORMDL2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>HYOU1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>BTG1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  is_on_pathway\n",
       "0     DICER1              1\n",
       "1       XPO5              1\n",
       "2     DROSHA              1\n",
       "3       AGO2              1\n",
       "4    ANKRD52              1\n",
       "495  ZSCAN22              0\n",
       "496  PCDHB16              0\n",
       "497   ORMDL2              0\n",
       "498    HYOU1              0\n",
       "499     BTG1              0"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([df_correlations[[\"gene\", \"is_on_pathway\"]][:5], df_correlations[[\"gene\", \"is_on_pathway\"]][-5:]])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
